cmd/zc_traverser_blob_versions.go

```diff
@@ -0,0 +1,130 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"github.com/Azure/azure-pipeline-go/pipeline"
+	"github.com/Azure/azure-storage-azcopy/common"
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"net/url"
+	"strings"
+)
+
+type blobVersionsTraverser struct {
+	rawURL                      *url.URL
+	p                           pipeline.Pipeline
+	ctx                         context.Context
+	recursive                   bool
+	includeDirectoryStubs       bool
+	incrementEnumerationCounter enumerationCounterFunc
+	listOfVersionIds            []string
+}
+
+func (t *blobVersionsTraverser) isDirectory(isSource bool) bool {
+	isDirDirect := copyHandlerUtil{}.urlIsContainerOrVirtualDirectory(t.rawURL)
+
+	// Skip the single blob check if we're checking a destination.
+	// This is an individual exception for blob because blob supports virtual directories and blobs sharing the same name.
+	if isDirDirect || !isSource {
+		return isDirDirect
+	}
+
+	_, err := t.getBlobProperties("")
+
+	return err != nil
+}
+
+func (t *blobVersionsTraverser) getBlobProperties(versionID string) (props *azblob.BlobGetPropertiesResponse, err error) {
+	blobURLParts := azblob.NewBlobURLParts(*t.rawURL)
+	blobURLParts.BlobName = strings.TrimSuffix(blobURLParts.BlobName, common.AZCOPY_PATH_SEPARATOR_STRING)
+	if versionID != "" {
+		blobURLParts.VersionID = versionID
+	}
+
+	blobURL := azblob.NewBlobURL(blobURLParts.URL(), t.p)
+	props, err = blobURL.GetProperties(t.ctx, azblob.BlobAccessConditions{})
+
+	if err != nil {
+		return props, err
+	}
+
+	// if there was no problem getting the properties, it implies that we are looking at a single blob
+	if gCopyUtil.doesBlobRepresentAFolder(props.NewMetadata()) {
+		return props, errors.New("This is not a blob")
+	}
+	return props, nil
+
+}
+
+func (t *blobVersionsTraverser) traverse(preprocessor objectMorpher, processor objectProcessor, filters []objectFilter) (err error) {
+	blobURLParts := azblob.NewBlobURLParts(*t.rawURL)
+
+	for _, versionID := range t.listOfVersionIds {
+		blobProperties, err := t.getBlobProperties(versionID)
```

Comment:
>blobProperties, err := t.getBlobProperties(versionID) [](start = 2, length = 53)

while getting the properties for every version is the technically right thing to do, but in both of our scenarios (download and remove) we do not need the properties, so I suggest to not get the properties for now, since it's a perf hit (1 i/o per version in a serial chain of gets). We should have a check somewhere to verify that if --list-of-versions is specified, the destination has to be local or trash.

Reply:
When we construct stored object, we also create a field called  blobProperties.ContentLength(). 
How do you suggest we get that value? If it's not mandatory it get this value, we can easily remove getProperties call.